### DISCUSSION TITLE
How to use Danger on GitLab CI/CD?

### DISCUSSION THREAD

**Discussion by ejstembler** [danger/danger#1441] [Q&A] [answered]
Anyone have success using Danger on GitLab.com's CI/CD?  I have this post on StackOverflow: [How to use Danger on GitLab CI/CD?](https://stackoverflow.com/questions/76412099/how-to-use-danger-on-gitlab-ci-cd).

---

GitLab CI/CD is already set up to run Danger, you have to add this to your `.gitlab-ci.yml` file:

```yaml
include:
  - project: 'gitlab-org/quality/pipeline-common'
    file: '/ci/danger-review.yml'
```

**Reference**: [/ci/danger-review.yml](https://gitlab.com/gitlab-org/quality/pipeline-common/-/blob/master/ci/danger-review.yml)

I'm also overriding the ruby image via:

```yaml
danger-review:
  image: ruby:3.2.2
```

My `Gemfile` references the `danger-gitlab` gem:

```Gemfile
# frozen_string_literal: true

source 'https://rubygems.org'
git_source(:github) { |repo| "https://github.com/#{repo}.git" }

ruby '3.2.2'

# ...

gem 'danger-gitlab', '~> 8.0.0'
```

You should be able to add a Ruby-flavored `Dangerfile` and it should work:

```Dangerfile
# frozen_string_literal: true

# ignore inline messages that are outside of the current diff
gitlab.dismiss_out_of_range_messages

message "Author @#{gitlab.mr_author}"

# Make it more obvious that a PR is a work in progress and shouldn't be merged yet
warn('This pull request is a Work in Progress and not ready to merge') if gitlab.mr_title.include? 'WIP'

# Large MR
warn('Big MR! Big changes, big things may happen! Check them.') if git.lines_of_code > 100

# Encourage contributors to write useful descriptions
warn('Please provide a Merge Request description') if gitlab.mr_body.length < 5

# If these are all empty something has gone wrong, better to raise it in a comment
if git.modified_files.empty? && git.added_files.empty? && git.deleted_files.empty?
  fail 'This PR has no changes at all, this is likely a developer issue.'
end

if git.deletions > git.insertions
  message '🗑 Code Cleanup!'
end
```

However, in the `danger-review` stage it cannot resolve the `gitlab` references:

```console
bundler: failed to load command: danger (/usr/local/bundle/bin/danger)
/usr/local/bundle/gems/danger-9.3.0/lib/danger/danger_core/dangerfile.rb:73:in `method_missing':  (Danger::DSLError)
[!] Invalid `Dangerfile` file: undefined local variable or method `gitlab' for #<Danger::Dangerfile:0x00007f2243d7dcc0 ...
```

Any ideas?

**Comment by orta** [comment, marked as answer]
I would look at how GitLab uses danger, https://gitlab.com/gitlab-org/gitlab as  I assume they use it - but I'm guessing that the env vars / url slug for detecting gitlab CI  are not correctly marking it as such

**Reply by ejstembler** [reply]
Good suggestion.

Apparently, their repo uses their own [gitlab-dangerfiles](https://gitlab.com/gitlab-org/ruby/gems/gitlab-dangerfiles) gem which also packages their own default plugins and rules.  (I believe I should be able to add my custom rules too.)  I did get it to work using that.

Dangerfile:
```Dangerfile
# frozen_string_literal: true

require 'gitlab-dangerfiles'

Gitlab::Dangerfiles.for_project(self) do |dangerfiles|
  # Imports all plugins, rules and the default reviewer roulette
  dangerfiles.import_defaults
end
```
Gemfile:
```Gemfile
# frozen_string_literal: true

source 'https://rubygems.org'
git_source(:github) { |repo| "https://github.com/#{repo}.git" }

ruby '3.2.2'

# ...

gem 'gitlab-dangerfiles', require: false
```
I also created an Access Token with the `Developer` role and `api` scope. And added the token to a new `DANGER_GITLAB_API_TOKEN` CI/CD variable.  This allowed the messages to show as comments on the MR.

![Screenshot_2023-06-06_at_12_05_55_PM](https://github.com/danger/danger/assets/45985338/93cd3ff7-0076-49d1-b078-65bd3b77438c)
